modules/gateway/gateway.go

```diff
@@ -36,58 +37,45 @@ type Gateway struct {
 	// network.
 	nodes map[modules.NetAddress]struct{}
 
-	// closeChan is used to shut down the Gateway's goroutines.
-	closeChan chan struct{}
+	// threads is used to signal the Gateway's goroutines to shut down and to wait
+	// for all goroutines to exit before returning from Close().
+	threads siasync.ThreadGroup
 
 	persistDir string
 
 	log *persist.Logger
-	mu  *sync.RWMutex
+	mu  sync.RWMutex
 }
 
 // Address returns the NetAddress of the Gateway.
 func (g *Gateway) Address() modules.NetAddress {
-	id := g.mu.RLock()
-	defer g.mu.RUnlock(id)
+	g.mu.RLock()
+	defer g.mu.RUnlock()
 	return g.myAddr
 }
```

Comment:
As an example. you don't need to use the ThreadGroup here because there's no use of a closeable resource in this thread.